### ISSUE TITLE
进行PaddleOCR识别训练时，内存不断泄漏，最终导致各种错误终止

### ISSUE THREAD

**Issue by swagzhang** [PaddlePaddle/PaddleOCR#3235]
情况一： num_workers设置超过1
在这种情况下，训练到一定次数后，突然训练进程就会被kill, 就只有被killed提示，无多余报错
![image](https://user-images.githubusercontent.com/14337160/124094726-9df01b80-da8b-11eb-8f25-1cc1abf55969.png)
情况二：num_workers设置为1
这种情况下不会直接被kill, 而是随着训练的进行，内存会被不断占用直至溢出报错，在过往的issue中看到要在reader_main.py中关掉多线程，但是当前最新版本已经没有这个文件了
这种情况会有不同错误，一个是之前说要alloc mem然后失败，还有一种是说shared memory不够。。。
![image](https://user-images.githubusercontent.com/14337160/124095496-51591000-da8c-11eb-8dd1-a3b4694cf577.png)
麻烦官方给予解决一下，感谢

**Comment by lz2470** [comment]
![image](https://user-images.githubusercontent.com/14337160/124102125-a9931080-da92-11eb-86a9-d2751cd7ae11.png)
又出现了这个报错，说打开了过多的file, 我一脸懵逼！

**Comment by LDOUBLEV** [comment]
首先清理你共享内存目录 /dev/shm

使用Dataloader 开启use_shared_memory 需要保证/dev/shm有8G以上的内存

如果遇到内存问题，建议把used_shared_memory 设置为False： https://github.com/PaddlePaddle/PaddleOCR/blob/2f4091e7163cf28f08b9e8ddc6772f6f89b72c60/configs/det/det_mv3_db.yml#L103

**Comment by lz2470** [comment]
我按照你说的在loader下面加了use_shared_memory：Fasle
![image](https://user-images.githubusercontent.com/14337160/124240097-fc310300-db4c-11eb-858f-c920c8e72fe0.png)
但是，我训练num_workers > 1还是在100来个epoch后莫名其妙的被killed，没有额外报错
![image](https://user-images.githubusercontent.com/14337160/124240256-271b5700-db4d-11eb-9d6e-b48cbc689efa.png)
我现在跑一下num_workers 为1的情况，我估计还是会因为内存溢出或者too many open files 然后挂掉。。。

**Comment by nuaa030710312** [comment]
官方什么时候能解决

**Comment by lz2470** [comment]
> 我按照你说的在loader下面加了use_shared_memory：Fasle
> ![image](https://user-images.githubusercontent.com/14337160/124240097-fc310300-db4c-11eb-858f-c920c8e72fe0.png)
> 但是，我训练num_workers > 1还是在100来个epoch后莫名其妙的被killed，没有额外报错
> ![image](https://user-images.githubusercontent.com/14337160/124240256-271b5700-db4d-11eb-9d6e-b48cbc689efa.png)
> 我现在跑一下num_workers 为1的情况，我估计还是会因为内存溢出或者too many open files 然后挂掉。。。

![image](https://user-images.githubusercontent.com/14337160/124245247-639d8180-db52-11eb-8830-7a965907da48.png)
跑完了，不出所料，报了一个OSError: [Errno 24] Too many open files的错误，这个要怎么办，求助

**Comment by lz2470** [comment]
![image](https://user-images.githubusercontent.com/14337160/124262542-c3048d00-db64-11eb-88ce-c24f705e80f9.png)
我把num_workers设置成0，然后batch size调到128，慢慢就会memory溢出了。。。
`Global:
  use_gpu: True
  epoch_num: 5000
  log_smooth_window: 20
  print_batch_step: 10
  save_model_dir: ./output/rec_en_number_lite
  save_epoch_step: 10
  # evaluation is run every 5000 iterations after the 4000th iteration
  eval_batch_step: [0, 200]
  # if pretrained_model is saved in static mode, load_static_weights must set to True
  cal_metric_during_train: True
  pretrained_model: ./output/rec_en_number_lite/latest
  checkpoints:
  save_inference_dir:
  use_visualdl: False
  infer_img:
  # for data or label process
  character_dict_path: ./ppocr/utils/en_dict.txt
  character_type: EN
  max_text_length: 250
  infer_mode: False
  use_space_char: True


Optimizer:
  name: Adam
  beta1: 0.9
  beta2: 0.999
  lr:
    name: Cosine
    learning_rate: 0.001
  regularizer:
    name: 'L2'
    factor: 0.00001

Architecture:
  model_type: rec
  algorithm: CRNN
  Transform:
  Backbone:
    name: MobileNetV3
    scale: 0.5
    model_name: small
    small_stride: [1, 2, 2, 2]
  Neck:
    name: SequenceEncoder
    encoder_type: rnn
    hidden_size: 48
  Head:
    name: CTCHead
    fc_decay: 0.00001

Loss:
  name: CTCLoss

PostProcess:
  name: CTCLabelDecode

Metric:
  name: RecMetric
  main_indicator: acc

Train:
  dataset:
    name: SimpleDataSet
    data_dir: ./train_data/
    label_file_list: ["./train_data/train_list.txt"]
    transforms:
      - DecodeImage: # load image
          img_mode: BGR
          channel_first: False
      - RecAug: 
      - CTCLabelEncode: # Class handling label
      - RecResizeImg:
          image_shape: [3, 32, 320]
      - KeepKeys:
          keep_keys: ['image', 'label', 'length'] # dataloader will return list in this order
  loader:
    shuffle: True
    batch_size_per_card: 128
    drop_last: True
    num_workers: 0
    use_shared_memory: False

Eval:
  dataset:
    name: SimpleDataSet
    data_dir: ./train_data/
    label_file_list: ["./train_data/eval_list.txt"]
    transforms:
      - DecodeImage: # load image
          img_mode: BGR
          channel_first: False
      - CTCLabelEncode: # Class handling label
      - RecResizeImg:
          image_shape: [3, 32, 320]
      - KeepKeys:
          keep_keys: ['image', 'label', 'length'] # dataloader will return list in this order
  loader:
    shuffle: False
    drop_last: False
    batch_size_per_card: 128
    num_workers: 0
    use_shared_memory: False`

**Comment by LDOUBLEV** [comment]
> 我把num_workers设置成0，然后batch size调到128，慢慢就会memory溢出了。。。

/dev/shm 目录下还有多少空间

**Comment by lz2470** [comment]
![image](https://user-images.githubusercontent.com/14337160/124406251-60cda700-dd73-11eb-9c14-072179276e3d.png)
有8个G...
但我感觉跟shared mem没关系吧，是我内存不断被侵蚀。。。说明代码里被alloc的memory没有在一次epoch后释放掉，有地方泄漏了吧，然后慢慢越攒越多直至崩溃？
![image](https://user-images.githubusercontent.com/14337160/124406312-9bcfda80-dd73-11eb-8305-332a9a80032c.png)
我16G的内存开始跑的时候差不多占用4，5G然后慢慢就满了。。

**Comment by lz2470** [comment]
![image](https://user-images.githubusercontent.com/14337160/124406815-f289e400-dd74-11eb-955b-3982bf0b137f.png)
跑了70个epoch就增加了2g内存
![image](https://user-images.githubusercontent.com/14337160/124407355-48ab5700-dd76-11eb-839d-9e26f95cfed7.png)
又增加了2g内存，而shm没有任何增加
![image](https://user-images.githubusercontent.com/14337160/124409719-6a5b0d00-dd7b-11eb-8788-926b79970f45.png)
 快满了。。
![image](https://user-images.githubusercontent.com/14337160/124410298-bce8f900-dd7c-11eb-918c-5f44adefa413.png)
 我这样短短续续训练了好几天，才估计只有3,4千个epoch。。。太累了，有没有办法不断掉的，让我顺利训练完。。。

**Comment by lz2470** [comment]
![image](https://user-images.githubusercontent.com/14337160/124529891-0b120100-de3e-11eb-8faa-7149a80ebc47.png)
最后就是内存跑没了。。。这我换成32G的也不够啊。。。迟早也会溢出如果内存不断泄漏的话

**Comment by neverstoplearn** [comment]
遇到同样的问题

**Comment by neverstoplearn** [comment]
请问有解决方案吗？

**Comment by MissPenguin** [comment]
拉取最新dygraph分支代码试试呢

**Comment by zjj421** [comment]
Release/2.1, 机器内存500多G,设置的共享内存250多G，同样训练文本检测模型ch_det_res18_db_v2.0.yml，每隔一段时间训练程序就会被killed，没有任何报错，官方可以在源码里的try...exception语句内打印出具体的报错信息吗？

**Comment by qiujian1232** [comment]
同这个问题，内存一直被侵占，应该在哪里改才能边训练边释放内存呢？
